piptools/repositories/pypi.py

```diff
@@ -282,21 +284,29 @@ def get_dependencies(self, ireq):
     def get_hashes(self, ireq):
         """
         Given an InstallRequirement, return a set of hashes that represent all
-        of the files for a given requirement. Editable requirements return an
+        of the files for a given requirement. Unhashable requirements return an
         empty set. Unpinned requirements raise a TypeError.
         """
-        if ireq.editable:
-            return set()
-
-        if is_url_requirement(ireq):
-            # url requirements may have been previously downloaded and cached
-            # locally by self.resolve_reqs()
-            cached_path = os.path.join(self._download_dir, ireq.link.filename)
-            if os.path.exists(cached_path):
-                cached_link = Link(path_to_url(cached_path))
-            else:
-                cached_link = ireq.link
-            return {self._get_file_hash(cached_link)}
+
+        if ireq.link:
+            link = ireq.link
+
+            if is_vcs_url(link) or (is_file_url(link) and is_dir_url(link)):
+                # Return empty set for unhashable requirements.
+                # Unhashable logic modeled on pip's
+                # RequirementPreparer.prepare_linked_requirement
```

Comment:
What about this block: https://github.com/pypa/pip/blob/master/src/pip/_internal/operations/prepare.py#L308-L316? Maybe it also should be included here?

Reply:
That block (`if not req.original_link and not req.is_pinned`) is covered by the logic later in the function. `if is_url_requirement(ireq)` will grab everything with ireq.original_link set, and `if not is_pinned_requirement(ireq)` will handle anything that isn't pinned.

(I compared pip-tools is_pinned_requirement vs. pip's req.is_pinned, and I think the only difference is pip-tools also checks for unpinned requirements like "==1.2.*")

Reply:
Right!